### PR TITLE
SD 1.5 Support For Advanced Lora Training (train_dreambooth_lora_sdxl_advanced.py)

### PR DESCRIPTION
**Pull request by brandostrong**
A quick adaptation to support people interested in using this method on 1.5 models.

# What does this PR do?

Adds support for 1.5 models on the advanced dreambooth training method laid out by @linoytsaban 

## Who can review?

@linoytsaban @sayakpaul

### CODE REVIEW

**Review by linoytsaban — CHANGES_REQUESTED** [review]
Thanks for the PR! looks great, excited to see what LoRAs will come out of the advanced script with SD!

I left a few comments regarding compatibility issues for the prompt text encoding, and unit added conditions. I think all should be quite quick to fix if you don't mind taking another look 😊
[examples/dreambooth/train_dreambooth_lora.py](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py)  - this might help with making those adjustments :)

**Comment by brandostrong** [comment]
Thank you so much for the comments Linoy! I appreciate how much detail you just provided me. Admittedly, this was something I quickly put together since I saw a demand for 1.5. It's rather late right now, so when I find time tomorrow I'll make the necessary changes.

Also, did you happen to test your script at all on 1.5, and do you have any comments on the efficacy of it?

**Comment by linoytsaban** [comment]
Awesome thanks so much😁
admittedly I only tried the advanced features with sdxl, so I'm intrigued!

**Comment by brandostrong** [comment]
I've made the suggested changes, and it runs but I'm not sure if it's training correctly. There seems to be some fundamental differences(at least in naming, this is a bit over my head) between the base sdxl and 1.5 training scripts, and in the case of encode_prompt we have fairly different parameters and arguments.  Could you give it a look? Is there an example dataset and **full** list of parameters that can be shared and compared as well?

**Comment by linoytsaban** [comment]
Thanks a lot! 

> There seems to be some fundamental differences(at least in naming, this is a bit over my head) between the base sdxl and 1.5 training scripts, but despite this it seems to work. Could you give it a test as well?

Sure! I'll go over it and run some experiments :)

**Comment by brandostrong** [comment]
Hi, I've spent the last few days testing and it holds up against my similar training in kohya. Both styles and likenesses are successful, though it really excels at styles.

**Comment by sayakpaul** [comment]
Thanks so much for your work!

Could we see some results? Also, let's get the code quality fixed, so that we can merge :)

**Comment by brandostrong** [comment]
Here are some results, I've attached the embedding and lora of a style(and somewhat character) train. 
I would say it trains just fine :)

[offscriptstyleadv.zip](https://github.com/huggingface/diffusers/files/14003279/offscriptstyleadv.zip)
The instance prompt is "in the style of offscriptstyle"

And attached are some example images from comfyui and wandb:
NVIDIA_Share_MrarE8lvp1
NVIDIA_Share_m5PQrQRVTq
image

**Comment by sayakpaul** [comment]
Cc: @linoytsaban

**Comment by linoytsaban** [comment]
Looks good @brandostrong! if you don't mind taking a look and merge my 2 suggestions (bug fixes for checkpoint saving) I think we'd be good to go 😁

**Comment by linoytsaban** [comment]
> Here are some results training a 2d style off the base sd 1.5 checkpoint, I've attached the embedding and lora of a style(and somewhat character) train. I would say it trains just fine :). @sayakpaul @linoytsaban , thanks for the help, hopefully this meets the quality requirements.
> 
> [offscriptstyleadv.zip](https://github.com/huggingface/diffusers/files/14003279/offscriptstyleadv.zip) The instance prompt is "in the style of offscriptstyle"
> 
> And attached are some example images from comfyui and wandb: NVIDIA_Share_MrarE8lvp1 NVIDIA_Share_m5PQrQRVTq image

btw @brandostrong - did you try prompting "in the style of `<s0><s1>` offscriptstyle"? Just curious because that way we also leverage the new optimized embeddings which should improve results even more :)

**Comment by brandostrong** [comment]
Ah oops, the entire essence of this training workflow, totally forgot. I'll send screenshots in an hour or so with the same seed/prompt. I'm running a long animatediff so my VRAM is currently parked at max

**Comment by brandostrong** [comment]
> Looks good @brandostrong! if you don't mind taking a look and merge my 2 suggestions (bug fixes for checkpoint saving) I think we'd be good to go 😁

Was the merge I just made the suggestions? I'm not that experienced with much more than solo git.

**Comment by linoytsaban** [comment]
> Ah oops, the entire essence of this training workflow, totally forgot. I'll send screenshots in an hour or so with the same seed/prompt. I'm running a long animatediff so my VRAM is currently parked at max

no worries! 😁

**Comment by linoytsaban** [comment]
> > Looks good @brandostrong! if you don't mind taking a look and merge my 2 suggestions (bug fixes for checkpoint saving) I think we'd be good to go 😁
> 
> Was the merge I just made the suggestions? I'm not that experienced with much more than solo git.

I believe they were not merged yet :) can you see my new comments? I think that would do it 😊

File: examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py [hunk 1]
```diff
@@ -0,0 +1,1954 @@
...
+                    raise ValueError(f"unexpected save model: {model.__class__}")
+
+                # make sure to pop weight so that corresponding model is not saved again
+                weights.pop()
+
+            StableDiffusionPipeline.save_lora_weights(
+                output_dir,
+                unet_lora_layers=unet_lora_layers_to_save,
+                text_encoder_lora_layers=text_encoder_one_lora_layers_to_save,
+            )
```
Comment (linoytsaban) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py:1280]:
```suggestion
            StableDiffusionPipeline.save_lora_weights(
                output_dir,
                unet_lora_layers=unet_lora_layers_to_save,
                text_encoder_lora_layers=text_encoder_one_lora_layers_to_save,
            )
        if args.train_text_encoder_ti:
            embedding_handler.save_embeddings(f"{output_dir}/{args.output_dir}_emb.safetensors")
```

File: examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py [hunk 2]
```diff
@@ -0,0 +1,1954 @@
...
+            text_encoder_one_lora_layers_to_save = None
+
+            for model in models:
+                if isinstance(model, type(accelerator.unwrap_model(unet))):
+                    unet_lora_layers_to_save = convert_state_dict_to_diffusers(get_peft_model_state_dict(model))
+                elif isinstance(model, type(accelerator.unwrap_model(text_encoder_one))):
+                    text_encoder_one_lora_layers_to_save = convert_state_dict_to_diffusers(
+                        get_peft_model_state_dict(model)
+                    )
+                else:
```
Comment (linoytsaban) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py:1270]:
```suggestion
                elif isinstance(model, type(accelerator.unwrap_model(text_encoder_one))):
                    if args.train_text_encoder:
                        text_encoder_one_lora_layers_to_save = convert_state_dict_to_diffusers(
                            get_peft_model_state_dict(model)
                        )
```

Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py:1270]: makes sure we save checkpoints correctly, as per bug fix in https://github.com/huggingface/diffusers/pull/6464

File: examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py (hunk 1 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py:1280]: makes sure we save checkpoints correctly, as per bug fix in https://github.com/huggingface/diffusers/pull/6464

**Review by linoytsaban — APPROVED** [review]

**Comment by linoytsaban** [comment]
@brandostrong I just committed 2 minor changes to `save_model_hook` - makes sure we save checkpoints correctly, as per the bug fix in #6464. I think we can ship now 😁🛳️

**Review by brandostrong — COMMENTED** [review]
Awesome, thank you :D
